File: stock_search/stats_families.py
"""Define shared stat-family groupings and timestamp fields."""

from __future__ import annotations

from typing import Literal

StatsFamily = Literal[
    "market_data",
    "market_snapshot",
    "statistics",
    "financials",
    "ratings",
]

STAT_FAMILIES: tuple[StatsFamily, ...] = (
    "market_data",
    "market_snapshot",
    "statistics",
    "financials",
    "ratings",
)

FAMILY_FIELDS: dict[StatsFamily, tuple[str, ...]] = {
    "market_data": (
        "price",
        "change",
        "change_percent_1d",
    ),
    "market_snapshot": (
        "name",
        "quote_type",
        "iv",
        "rsi",
        "change_percent_1m",
        "change_percent_3m",
        "change_percent_6m",
        "change_percent_1y",
        "change_percent_mtd",
        "change_percent_ytd",
    ),
    "statistics": (
        "market_cap",
        "pe",
        "pe_forward",
        "peg",
        "beta",
        "free_cash_flow",
    ),
    "financials": (
        "revenue_growth",
        "gross_margin",
        "debt_to_equity",
    ),
    "ratings": (
        "median_upside",
        "ratings",
    ),
}

FIELD_TO_FAMILY: dict[str, StatsFamily] = {field: family for family, fields in FAMILY_FIELDS.items() for field in fields}

FAMILY_TIMESTAMP_FIELD: dict[StatsFamily, str] = {
    "market_data": "market_data_fetched_at",
    "market_snapshot": "market_snapshot_fetched_at",
    "statistics": "statistics_fetched_at",
    "financials": "financials_fetched_at",
    "ratings": "ratings_fetched_at",
}

BLOCKING_AUTO_FAMILIES: frozenset[StatsFamily] = frozenset({"market_data"})
NON_BLOCKING_AUTO_FAMILIES: frozenset[StatsFamily] = frozenset(family for family in STAT_FAMILIES if family not in BLOCKING_AUTO_FAMILIES)


def family_timestamp_fields(family: StatsFamily) -> tuple[str, ...]:
    """Return timestamp fields accepted for one stat family."""
    return (FAMILY_TIMESTAMP_FIELD[family],)

File: src/stock-search/stats-families.ts
/** Shared stat-family groupings and cache windows for the TypeScript resolver. */

export type StatsFamily =
	| "market_data"
	| "market_snapshot"
	| "statistics"
	| "financials"
	| "ratings";

export const STAT_FAMILIES: StatsFamily[] = [
	"market_data",
	"market_snapshot",
	"statistics",
	"financials",
	"ratings",
];

export const FAMILY_FIELDS: Record<StatsFamily, string[]> = {
	market_data: ["price", "change", "change_percent_1d"],
	market_snapshot: [
		"name",
		"quote_type",
		"sector_name",
		"industry_name",
		"iv",
		"change_percent_1m",
		"change_percent_3m",
		"change_percent_6m",
		"change_percent_1y",
	],
	statistics: [
		"market_cap",
		"fx",
		"revenue",
		"pe",
		"pe_forward",
		"ps",
		"ps_forward",
		"peg",
		"beta",
		"roe",
		"roic",
		"debt_to_equity",
		"free_cash_flow",
		"shareholder_yield",
		"rsi",
	],
	financials: [
		"revenue",
		"revenue_growth",
		"revenue_growth_1y",
		"revenue_cagr_3y",
		"eps_growth",
		"fcf_growth_1y",
		"fcf_cagr_3y",
		"gross_margin",
		"gross_margin_median_3y",
		"operating_margin",
		"operating_margin_median_3y",
		"operating_margin_delta_vs_3y",
		"operating_margin_std_3y",
		"free_cash_flow",
		"fcf_margin_median_3y",
		"shares_change_1y",
		"shares_change_cagr_3y",
		"financials_currency",
		"research_and_development",
		"rd_intensity",
		"rd_knowledge_capital",
	],
	ratings: ["median_upside", "ratings"],
};

export const FAMILY_TIMESTAMP_FIELD: Record<StatsFamily, string> = {
	market_data: "market_data_fetched_at",
	market_snapshot: "market_snapshot_fetched_at",
	statistics: "statistics_fetched_at",
	financials: "financials_fetched_at",
	ratings: "ratings_fetched_at",
};

export const BLOCKING_AUTO_FAMILIES = new Set<StatsFamily>(["market_data"]);

export const FAMILY_POLICIES: Record<
	StatsFamily,
	{ freshWindowMs: number; staleWindowMs: number; failureCooldownMs: number }
> = {
	market_data: {
		freshWindowMs: 60_000,
		staleWindowMs: 600_000,
		failureCooldownMs: 180_000,
	},
	market_snapshot: {
		freshWindowMs: 3_600_000,
		staleWindowMs: 172_800_000,
		failureCooldownMs: 1_800_000,
	},
	statistics: {
		freshWindowMs: 86_400_000,
		staleWindowMs: 172_800_000,
		failureCooldownMs: 1_800_000,
	},
	financials: {
		freshWindowMs: 86_400_000,
		staleWindowMs: 172_800_000,
		failureCooldownMs: 1_800_000,
	},
	ratings: {
		freshWindowMs: 86_400_000,
		staleWindowMs: 172_800_000,
		failureCooldownMs: 1_800_000,
	},
};
